File: SER423/Android/app/src/main/java/edu/asu/bsse/gcarvaj3/ser423labapp/PlaceLibrary.java
package edu.asu.bsse.gcarvaj3.ser423labapp;

import org.json.JSONException;
import org.json.JSONObject;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Copyright 2019 Gianni Carvajal
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author Gianni Carvajal mailto:gcarvaj3@asu.edu
 * @version November 25, 2019
 */

public class PlaceLibrary {
    private ArrayList<PlaceDescription> placeLibrary;
    private static PlaceLibrary library = null;

    private PlaceLibrary(String fileName) {
        placeLibrary = readFromJson(fileName);
    }
    private PlaceLibrary() {
        placeLibrary = new ArrayList<>();
    }

    public static PlaceLibrary getInstance(String fileName) {
        if (library == null) {
            if (fileName.contains("{")) library = new PlaceLibrary(fileName);
            else {
                library = new PlaceLibrary();
            }
        }
        return library;
    }


    private ArrayList<PlaceDescription> readFromJson(String fileLocation) {
        try {
            ArrayList<PlaceDescription> places = new ArrayList<>();
            JSONObject placeLibraryJSON = new JSONObject(fileLocation);
            Iterator<String> keys = placeLibraryJSON.keys();
            while (keys.hasNext()){
                String key = keys.next();
                if (placeLibraryJSON.get(key) instanceof JSONObject) {
                    places.add(new PlaceDescription(placeLibraryJSON.get(key).toString()));
                }
            }
            return places;
        } catch (JSONException e) {
            android.util.Log.e(this.getClass().getSimpleName(),
                    "error converting to/from json from file");
            android.util.Log.e(this.getClass().getSimpleName(), e.toString());
            return null;
        }

    }

    public String toJSONString() {
        if (placeLibrary == null) {
            return "No library found";
        }
        String jsonStrToReturn = "{";
        for (int i = 0; i < placeLibrary.size(); i++) {
            jsonStrToReturn += placeLibrary.get(i).getName() + ": " +
                    placeLibrary.get(i).toJSONString();
        }
        jsonStrToReturn += "}";
        return jsonStrToReturn;
    }

    public ArrayList<PlaceDescription> getPlaceLibrary() {
        return placeLibrary;
    }

    public boolean findName(PlaceLibrary library, String name) {
        AtomicReference<Boolean> found = new AtomicReference<>(false);
        if (library.getPlaceLibrary().isEmpty()) return false;
        library.getPlaceLibrary().forEach(location -> {
            if (location.getName().equals(name)) {
                found.set(true);
            }
        });
        return found.get();
    }
}

File: SER423/iOS/SER423Lab/SER423Lab/PlaceLibrary.swift
/**
* Copyright 2019 Gianni Carvajal
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
* @author Gianni Carvajal mailto:gcarvaj3@asu.edu
* @version November 28, 2019
*/

import Foundation

final class PlaceLibrary {
    
    static let sharedLibrary = PlaceLibrary()
    private var library = [PlaceDescription]()
    
    private init() {}
    
    func getLibrary() -> [PlaceDescription] {
        return PlaceLibrary.sharedLibrary.library
    }
    
    func addPlaceToLibrary(placeToAdd: PlaceDescription) {
        PlaceLibrary.sharedLibrary.library.append(placeToAdd)
    }
    
    func removePlace(placeToRemove: PlaceDescription) {
        let index = find(value: placeToRemove.name)
        print(index)
        if index != nil {
            PlaceLibrary.sharedLibrary.library.remove(at: index!)
        }
        else {
            NSLog("Location not in library. Not removing.")
        }
    }
    
    func find(value searchValue: String) -> Int?
    {
        for (index, value) in PlaceLibrary.sharedLibrary.getLibrary().enumerated()
        {
            if value.name == searchValue {
                return index
            }
        }
    
        return nil
    }
    
    func updateLocation(placeToUpdate: PlaceDescription) {
        let index = find(value: placeToUpdate.name)
        if index != nil {
            PlaceLibrary.sharedLibrary.library[index!] = placeToUpdate
        }
    }
}
